File: python/05.py
#! /usr/bin/env python3

rules = []
page_lists = []

with open('../inputs/input05.txt') as fp:
    for line_str in fp.readlines():
        line = line_str.strip()
        if '|' in line:
            parts = [int(part) for part in line.split('|')]
            rules.append(parts)
        elif ',' in line:
            page_lists.append([int(part) for part in line.split(',')])


# Check if a single rule is passed by its members positions in a page list
def is_rule_passed(rule, page_list):
    if rule[0] not in page_list or rule[1] not in page_list:
        return True
    index_a = page_list.index(rule[0])
    index_b = page_list.index(rule[1])
    return index_a < index_b


# Check if all rules are passed
def is_all_rules_passed(page_list):
    return all([is_rule_passed(rule, page_list) for rule in rules])


# Get the element in the middle of a list
def middle_of_list(page_list):
    return page_list[len(page_list) // 2]


##
# Part 1
# Sum the middle elements of all valid page lists
#
valid_page_lists = [page_list for page_list in page_lists if is_all_rules_passed(page_list)]
middle_pages = [middle_of_list(page_list) for page_list in valid_page_lists]
print("Part 1:", sum(middle_pages))


# find the probable middle element, based on finding the element with an equal number of siblings before and after
def find_middle(page_list):
    for n in page_list:
        must_come_before = [rule[0] for rule in rules if rule[1] == n and rule[0] in page_list]
        must_come_after = [rule[1] for rule in rules if rule[0] == n and rule[1] in page_list]
        if len(must_come_before) == len(must_come_after):
            return n
    return None


##
# Part 2
# Sum the middle elements of all invalid page lists
#
invalid_page_lists = [page_list for page_list in page_lists if not is_all_rules_passed(page_list)]
middle_pages = [find_middle(page_list) for page_list in invalid_page_lists]
print("Part 2:", sum(middle_pages))


# P1: 4957
# P2: 6938

File: rust/aoc/src/bin/05.rs
use std::fs::read_to_string;

fn read_lines(filename: &str) -> (Vec<Vec<u32>>, Vec<Vec<u32>>) {
    let mut rules = Vec::new();
    let mut page_lists = Vec::new();

    read_to_string(filename)
        .unwrap()
        .lines()
        .for_each(|line| {
            let line_str = line.trim();

            if line_str.find('|').is_some() {
                let parts = line_str.split('|').map(|part| part.parse::<u32>().unwrap()).collect::<Vec<u32>>();
                rules.push(parts)
            }
            else if line_str.find(',').is_some() {
                let parts = line_str.split(',').map(|part| part.parse::<u32>().unwrap()).collect::<Vec<u32>>();
                page_lists.push(parts)
            }
        });

    (rules, page_lists)
}

// Check if a single rule is passed by its members positions in a page list
fn is_rule_passed(rule: Vec<u32>, page_list: Vec<u32>) -> bool {
    if !page_list.contains(&rule[0]) || !page_list.contains(&rule[1]) {
        return true
    }
    let index_a = page_list.iter().position(|r| r == &rule[0]); // can be None or Some(isize)
    let index_b = page_list.iter().position(|r| r == &rule[1]);
    index_a < index_b
}

// Check if all rules are passed
fn is_all_rules_passed(rules: Vec<Vec<u32>>, page_list: Vec<u32>) -> bool {
    rules.iter().all(|rule| is_rule_passed(rule.to_vec(), page_list.clone()))
}

// Get the element in the middle of a list
fn middle_of_list(page_list: Vec<u32>) -> u32 {
    let len = page_list.len() as f32;
    let mid = (len / 2.0).floor() as usize;
    page_list[mid]
}

// Find the probable middle element, based on finding the element with an equal number of siblings before and after
fn find_middle(rules: Vec<Vec<u32>>, page_list: Vec<u32>) -> u32 {
    let mut middle: u32 = 0;
    page_list
    .iter()
    .any(|n| {
        let mut must_come_before = 0;
        let mut must_come_after = 0;
        rules.iter().for_each(|rule| {
            if &rule[1] == n && page_list.contains(&rule[0]) {
                must_come_before += 1
            }
            if &rule[0] == n && page_list.contains(&rule[1]) {
                must_come_after += 1
            }
        });
        if must_come_before == must_come_after {
            middle = *n;
            true
        }
        else {
            false
        }
    });
    middle
}


// Part 1
// Sum the middle elements of all valid page lists
fn part1(rules: Vec<Vec<u32>>, page_lists: Vec<Vec<u32>>) -> u32 {
    let valid_page_lists = page_lists.iter().filter(|page_list| is_all_rules_passed(rules.clone(), page_list.to_vec()));
    let middle_pages = valid_page_lists.map(|page_list| middle_of_list(page_list.to_vec()));
    middle_pages.sum()
}

// Part 2
// Sum the middle elements of all invalid page lists
fn part2(rules: Vec<Vec<u32>>, page_lists: Vec<Vec<u32>>) -> u32 {
    let invalid_page_lists = page_lists.iter().filter(|page_list| !is_all_rules_passed(rules.clone(), page_list.to_vec()));
    let middle_pages = invalid_page_lists.map(|page_list| find_middle(rules.clone(), page_list.to_vec()));
    middle_pages.sum()
}

fn main() {
    let (rules, page_lists) = read_lines("../../inputs/input05.txt");
    println!("P1: {}", part1(rules.clone(), page_lists.clone()));
    println!("P2: {}", part2(rules.clone(), page_lists.clone()));
}

// P1: 4957
// P2: 6938
